src/libPMacc/include/eventSystem/events/EventPool.hpp

```diff
@@ -23,79 +23,119 @@
-         * destroys all cuda events in the pool
+
+        /** add CudaEvent to the pool
+         *
+         * the pool takes the ownership of the pointer
+         *
+         * @param ev pointer to CudaEvent
          */
-        virtual ~EventPool()
+        void push( CudaEvent* const ev )
         {
-            for (size_t i = 0; i < events.size(); i++)
+            /* Guard that no event is added during the pool is closed (shutdown phase).
+             * This function is also called during the destructor is evaluated.
+             */
+            if( !isClosed )
+                freeEvents.push_back( CudaEventHandle(ev) );
+        }
+
+        /** create and add cuda events to the pool
+         *
+         * @param count number of cuda events to add
+         */
+        void createEvents( size_t count = 1u )
+        {
+            for( size_t i = 0u; i < count; i++ )
             {
-                log(ggLog::CUDA_RT()+ggLog::MEMORY(),"Sync and Delete Event: %1%") % i;
-                CudaEvent::destroy(events[i]);
+                CudaEvent* nativeEvent = new CudaEvent( );
+                events.push_back( nativeEvent );
+                push( nativeEvent );
             }
         }
 
-        /**
-         * Returns the next cuda event in the event pool.
-         * @return the next cuda event
+        /** Returns the number of cuda events in the pool.
+         *
+         * @return number of cuda events
          */
-        CudaEvent getNextEvent()
+        size_t getEventsCount( )
         {
-            CudaEvent result = events[currentEventIndex];
-            currentEventIndex = (currentEventIndex + 1) % events.size();
-            return result;
+            return events.size( );
+        }
+
+    private:
+        friend class Environment< DIM1 >;
+        friend class Environment< DIM2 >;
+        friend class Environment< DIM3 >;
+
+        static EventPool& getInstance( )
+        {
+            static EventPool instance;
+            return instance;
         }
 
         /**
-         * Adds count new cuda events to the pool.
-         * @param count number of cuda events to add
+         * Constructor.
+         * adds a single cuda event to the pool
          */
-        void addEvents(size_t count)
+        EventPool( ) : isClosed( false )
         {
-            for (size_t i = 0; i < count; i++)
-            {
-                events.push_back(CudaEvent::create());
-            }
         }
 
         /**
-         * Returns the number of cuda events in the pool.
-         * @return number of cuda events
+         * Destructor.
+         * destroys all cuda events in the pool
          */
-        size_t getEventsCount()
+        ~EventPool()
         {
-            return events.size();
+            log( ggLog::CUDA_RT( )+ggLog::MEMORY( ), "shutdown EventPool with %1% events" ) % getEventsCount( );
+            isClosed = true;
+            freeEvents.clear( );
+            events.clear( );
```

Comment:
careful, there is no `delete` called for your members: https://stackoverflow.com/questions/22709344/does-stdvectortclear-call-destructor-of-the-content